File: src/position_move_test.cpp
/**
 * Position control test over CAN using SimpleFOC angle mode.
 *
 * Architecture:
 *   STM32 (full cascade): Position P → Velocity PID → Voltage (SimpleFOC)
 *   This program: sends position targets over CAN, monitors feedback.
 *
 * SimpleFOC handles all the control internally at full sensor rate (~150Hz AS5600).
 * We just tell it where to go.
 *
 * Usage: ./position_move_test [options]
 *   --vel-limit <rad/s>   Velocity limit sent to motor (default 10.0)
 *   --move-deg <degrees>  Move distance (default 360)
 *   --cycles <n>          Back-and-forth cycles (default 1)
 *   --settle <seconds>    Settle time at each end (default 1.5)
 *   -i <interface>        CAN interface (default can0)
 */

#include "can_interface.hpp"
#include "motor_protocol.hpp"

#include <chrono>
#include <cmath>
#include <csignal>
#include <cstring>
#include <iostream>
#include <string>
#include <thread>

volatile sig_atomic_t running = 1;
void signalHandler(int) { running = 0; }

static uint64_t nowMs() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}

static double nowSec() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count() / 1e6;
}

struct Config {
    std::string can_interface = "can0";
    float vel_limit = 10.0f;    // rad/s — sent to motor as velocity_limit
    float move_deg = 360.0f;
    float settle_time = 1.5f;
    int cycles = 1;
};

static Config parseArgs(int argc, char* argv[]) {
    Config cfg;
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--vel-limit" && i + 1 < argc) cfg.vel_limit = std::stof(argv[++i]);
        else if (arg == "--move-deg" && i + 1 < argc) cfg.move_deg = std::stof(argv[++i]);
        else if (arg == "--cycles" && i + 1 < argc) cfg.cycles = std::stoi(argv[++i]);
        else if (arg == "--settle" && i + 1 < argc) cfg.settle_time = std::stof(argv[++i]);
        else if (arg == "-i" && i + 1 < argc) cfg.can_interface = argv[++i];
        else if (arg == "-h" || arg == "--help") {
            std::cout << "Usage: " << argv[0] << " [options]\n"
                      << "  --vel-limit <rad/s>   Velocity limit (default 10.0)\n"
                      << "  --move-deg <degrees>  Move distance (default 360)\n"
                      << "  --cycles <n>          Back-and-forth cycles (default 1)\n"
                      << "  --settle <seconds>    Settle time (default 1.5)\n"
                      << "  -i <interface>        CAN interface (default can0)\n";
            exit(0);
        }
    }
    return cfg;
}

// Send position target + velocity limit over CAN
static void sendPositionTarget(CANInterface& can, float pos_rad, float vel_limit) {
    uint8_t data[8];
    std::memcpy(data, &pos_rad, 4);
    std::memcpy(data + 4, &vel_limit, 4);
    can.send(CAN_ID_PDO2_TX, data, 8);
}

// Drain CAN frames, return latest feedback
static bool drainAndGetLatest(CANInterface& can, MotorProtocol& motor,
                               float& pos_out, float& vel_out) {
    bool got = false;
    for (int i = 0; i < 50; i++) {
        auto frame = can.receive(i == 0 ? 1 : 0);
        if (!frame) break;
        if (motor.parseFrame(*frame)) {
            pos_out = motor.getState().position_rad;
            vel_out = motor.getState().velocity_rad_s;
            got = true;
        }
    }
    return got;
}

int main(int argc, char* argv[]) {
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);

    Config cfg = parseArgs(argc, argv);
    float move_rad = cfg.move_deg * M_PI / 180.0f;

    std::cout << "=== Position Move Test (SimpleFOC Angle Mode) ===" << std::endl;
    std::cout << "  Move: " << cfg.move_deg << " deg (" << move_rad << " rad)" << std::endl;
    std::cout << "  Vel limit: " << cfg.vel_limit << " rad/s" << std::endl;
    std::cout << "  Cycles: " << cfg.cycles << std::endl;
    std::cout << std::endl;

    CANInterface can;
    if (!can.open(cfg.can_interface)) {
        std::cerr << "Failed to open CAN interface" << std::endl;
        return 1;
    }
    MotorProtocol motor(can);

    // Wake up with heartbeats
    std::cout << "Connecting to motor..." << std::endl;
    for (int i = 0; i < 10; i++) {
        motor.sendHeartbeat();
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

    // Read initial position
    float pos = 0, vel = 0;
    bool got_pos = false;
    for (int i = 0; i < 20 && !got_pos; i++) {
        motor.sendHeartbeat();
        got_pos = drainAndGetLatest(can, motor, pos, vel);
        if (!got_pos) std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }
    if (!got_pos) {
        std::cerr << "ERROR: No feedback. Check CAN." << std::endl;
        return 1;
    }

    float pos_start = pos;
    std::cout << "  Start position: " << (pos_start * 180.0 / M_PI) << " deg" << std::endl;

    // Enable position mode (NMT MODE_POSITION = 0x13)
    std::cout << "Enabling position mode..." << std::endl;
    sendPositionTarget(can, pos_start, cfg.vel_limit);  // Hold current position
    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    motor.setMode(0x13);  // MODE_POSITION
    std::this_thread::sleep_for(std::chrono::milliseconds(300));
    motor.sendHeartbeat();
    std::cout << "  Motor enabled in angle mode" << std::endl << std::endl;

    // Build moves
    struct Move { std::string name; float target; };
    std::vector<Move> moves;
    for (int i = 0; i < cfg.cycles; i++) {
        moves.push_back({"forward", pos_start + move_rad});
        moves.push_back({"reverse", pos_start});
    }

    printf("%5s  %9s  %9s  %8s  %8s\n", "time", "target", "actual", "error", "vel");
    printf("---------------------------------------------------\n");

    uint64_t last_hb_ms = 0;
    double last_print_time = 0;

    try {
        for (auto& move : moves) {
            if (!running) break;

            float pos_target = move.target;
            printf("\n--- %s: target = %.1f deg ---\n",
                   move.name.c_str(), pos_target * 180.0 / M_PI);

            // Send position target to motor
            sendPositionTarget(can, pos_target, cfg.vel_limit);

            double move_start = nowSec();
            double settle_start = -1;
            bool settled = false;

            while (running && !settled) {
                uint64_t now_ms = nowMs();
                double now = nowSec();

                // Heartbeat every 80ms
                if (now_ms - last_hb_ms >= 80) {
                    motor.sendHeartbeat();
                    last_hb_ms = now_ms;
                }

                // Resend position target every 200ms (keep watchdog alive)
                static double last_cmd_time = 0;
                if (now - last_cmd_time >= 0.2) {
                    sendPositionTarget(can, pos_target, cfg.vel_limit);
                    last_cmd_time = now;
                }

                // Read feedback
                bool got = drainAndGetLatest(can, motor, pos, vel);
                if (!got) {
                    auto frame = can.receive(5);
                    if (frame) {
                        motor.parseFrame(*frame);
                        pos = motor.getState().position_rad;
                        vel = motor.getState().velocity_rad_s;
                        got = true;
                    }
                }
                if (!got) continue;

                float pos_error = pos_target - pos;

                // Print at ~4Hz
                if (now - last_print_time >= 0.25) {
                    printf("%5.1fs  %+8.1fd  %+8.1fd  %+7.1fd  %+7.2fr/s\n",
                           now - move_start,
                           pos_target * 180.0 / M_PI,
                           pos * 180.0 / M_PI,
                           pos_error * 180.0 / M_PI,
                           vel);
                    fflush(stdout);
                    last_print_time = now;
                }

                // Settled: within 3 deg and velocity under 0.5 rad/s
                if (std::fabs(pos_error) < (3.0 * M_PI / 180.0) &&
                    std::fabs(vel) < 0.5f) {
                    if (settle_start < 0) settle_start = now;
                    else if (now - settle_start >= cfg.settle_time) settled = true;
                } else {
                    settle_start = -1;
                }

                // Timeout
                if (now - move_start > 15) {
                    printf("  TIMEOUT at %.1f deg (error: %+.1f deg)\n",
                           pos * 180.0 / M_PI,
                           pos_error * 180.0 / M_PI);
                    break;
                }
            }

            if (settled) {
                float err = pos_target - pos;
                printf("  Settled at %.1f deg (error: %+.1f deg)\n",
                       pos * 180.0 / M_PI, err * 180.0 / M_PI);
            }
        }
    } catch (...) {}

    // Stop
    std::cout << "\nStopping motor..." << std::endl;
    motor.setMode(0x00);  // DISABLED
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    motor.sendHeartbeat();
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    if (drainAndGetLatest(can, motor, pos, vel)) {
        printf("Final position: %.1f deg\n", pos * 180.0 / M_PI);
    }

    std::cout << "Done." << std::endl;
    return 0;
}

File: scripts/position_move_test.py
#!/usr/bin/env python3
"""
Two-stage position control test over CAN.

Architecture:
  - STM32 (inner loop): Velocity PID via SimpleFOC (~10kHz)
  - This script (outer loop): Position P-controller over CAN (~50Hz)
    Reads AS5600 position feedback, outputs velocity commands.

The AS5600 is too slow for direct position PID on the motor side,
but works well for a host-side position loop that generates smooth
velocity setpoints for the fast inner loop.

Usage: python3 position_move_test.py [--vel-limit 5.0] [--move-deg 360]
"""

import socket
import struct
import time
import math
import sys
import argparse

# CAN config
CAN_INTERFACE = "can0"
DEVICE_ID = 1

# CAN IDs (match firmware)
CAN_ID_NMT       = 0x001
CAN_ID_PDO2_TX   = 0x301   # Host -> motor: [velocity_f32, unused_f32]
CAN_ID_PDO2_RX   = 0x281   # Motor -> host: [position_f32, velocity_f32]
CAN_ID_HEARTBEAT = 0x701

# Motor modes
MODE_DISABLED = 0x00
MODE_VELOCITY = 0x12


class CANBus:
    def __init__(self, interface):
        self.sock = socket.socket(socket.PF_CAN, socket.SOCK_RAW, socket.CAN_RAW)
        self.sock.bind((interface,))

    def send(self, can_id, data=b""):
        dlc = len(data)
        self.sock.send(struct.pack("<IB3x8s", can_id, dlc, data.ljust(8, b'\x00')))

    def recv(self, timeout_s=0.1):
        self.sock.settimeout(timeout_s)
        try:
            raw = self.sock.recv(16)
            can_id, dlc = struct.unpack("<IB3x", raw[:8])
            return (can_id & 0x7FF, raw[8:8 + dlc])
        except socket.timeout:
            return None

    def drain(self):
        self.sock.settimeout(0)
        try:
            while True:
                self.sock.recv(16)
        except (BlockingIOError, socket.timeout):
            pass

    def send_heartbeat(self):
        self.send(CAN_ID_HEARTBEAT, b'')

    def send_nmt(self, mode):
        self.send(CAN_ID_NMT, struct.pack('BB', mode, DEVICE_ID))

    def send_velocity(self, vel_rad_s):
        self.send(CAN_ID_PDO2_TX, struct.pack('<ff', vel_rad_s, 0.0))

    def close(self):
        self.sock.close()


def read_feedback(can, timeout_s=0.05):
    """Read one feedback frame. Returns (position, velocity) or None."""
    result = can.recv(timeout_s)
    if result is None:
        return None
    can_id, data = result
    if can_id == CAN_ID_PDO2_RX and len(data) >= 8:
        pos = struct.unpack('<f', data[0:4])[0]
        vel = struct.unpack('<f', data[4:8])[0]
        return (pos, vel)
    # Got a different frame (heartbeat echo, etc), not feedback
    return None


def drain_and_get_latest(can):
    """Drain all pending frames and return the latest feedback."""
    latest = None
    while True:
        fb = read_feedback(can, timeout_s=0)
        if fb is None:
            break
        latest = fb
    return latest


def main():
    parser = argparse.ArgumentParser(description="Position move test over CAN")
    parser.add_argument('--vel-limit', type=float, default=5.0,
                        help='Max velocity command (rad/s), default 5.0')
    parser.add_argument('--move-deg', type=float, default=360.0,
                        help='Move distance in degrees, default 360')
    parser.add_argument('--kp', type=float, default=2.0,
                        help='Position P gain (vel_cmd = Kp * pos_error), default 2.0')
    parser.add_argument('--settle-time', type=float, default=2.0,
                        help='Time to wait at each end (seconds), default 2.0')
    parser.add_argument('--cycles', type=int, default=1,
                        help='Number of back-and-forth cycles, default 1')
    parser.add_argument('--accel-limit', type=float, default=10.0,
                        help='Max acceleration (rad/s^2), default 10.0')
    args = parser.parse_args()

    move_rad = math.radians(args.move_deg)
    vel_limit = args.vel_limit
    kp = args.kp
    accel_limit = args.accel_limit

    print(f"=== Position Move Test ===")
    print(f"  Move: {args.move_deg:.0f} deg ({move_rad:.2f} rad)")
    print(f"  Vel limit: {vel_limit:.1f} rad/s")
    print(f"  Kp: {kp:.1f}")
    print(f"  Accel limit: {accel_limit:.1f} rad/s^2")
    print(f"  Cycles: {args.cycles}")
    print()

    # Connect to CAN
    can = CANBus(CAN_INTERFACE)

    # Activate motor: send heartbeats, then enable velocity mode
    print("Connecting to motor...")
    for _ in range(10):
        can.send_heartbeat()
        time.sleep(0.05)

    can.drain()

    # Read initial position
    print("Reading initial position...")
    can.send_heartbeat()
    time.sleep(0.1)
    pos_start = None
    for _ in range(20):
        fb = read_feedback(can, timeout_s=0.1)
        if fb is not None:
            pos_start = fb[0]
            break
        can.send_heartbeat()

    if pos_start is None:
        print("ERROR: No feedback from motor. Check CAN connection.")
        can.close()
        return 1

    print(f"  Start position: {math.degrees(pos_start):.1f} deg ({pos_start:.3f} rad)")

    # Enable velocity mode
    print("Enabling velocity mode...")
    can.send_velocity(0.0)  # Zero velocity first
    time.sleep(0.05)
    can.send_nmt(MODE_VELOCITY)
    time.sleep(0.3)
    can.send_heartbeat()
    print("  Motor enabled in velocity mode")
    print()

    # Position control loop
    current_vel_cmd = 0.0
    pos_target = pos_start  # Start at current position

    # Build list of moves: forward, wait, back, wait (repeated)
    moves = []
    for _ in range(args.cycles):
        moves.append(("forward", pos_start + move_rad))
        moves.append(("reverse", pos_start))

    last_hb_time = 0
    last_print_time = 0

    print(f"{'time':>5s}  {'target':>9s}  {'actual':>9s}  {'error':>8s}  {'vel_cmd':>8s}  {'vel_act':>8s}")
    print("-" * 62)

    try:
        for move_name, target in moves:
            pos_target = target
            print(f"\n--- {move_name}: target = {math.degrees(pos_target):.1f} deg ---")

            move_start = time.time()
            settled = False
            settle_start = None

            while True:
                now = time.time()

                # Send heartbeat every 80ms
                if now - last_hb_time >= 0.08:
                    can.send_heartbeat()
                    last_hb_time = now

                # Get latest feedback (drain queue)
                latest = drain_and_get_latest(can)
                if latest is None:
                    # Try blocking read
                    latest = read_feedback(can, timeout_s=0.02)
                if latest is None:
                    continue

                pos_actual, vel_actual = latest

                # Position P controller with velocity limiting
                pos_error = pos_target - pos_actual
                vel_desired = kp * pos_error

                # Clamp velocity to limit
                vel_desired = max(-vel_limit, min(vel_limit, vel_desired))

                # Rate-limit acceleration
                dt = 0.02  # ~50Hz feedback rate
                max_dv = accel_limit * dt
                dv = vel_desired - current_vel_cmd
                if abs(dv) > max_dv:
                    dv = max_dv if dv > 0 else -max_dv
                current_vel_cmd += dv

                # Send velocity command
                can.send_velocity(current_vel_cmd)

                # Print status at ~4Hz
                if now - last_print_time >= 0.25:
                    print(f"{now - move_start:5.1f}s  "
                          f"{math.degrees(pos_target):+8.1f}d  "
                          f"{math.degrees(pos_actual):+8.1f}d  "
                          f"{math.degrees(pos_error):+7.1f}d  "
                          f"{current_vel_cmd:+7.2f}r/s  "
                          f"{vel_actual:+7.2f}r/s")
                    last_print_time = now

                # Check if settled (within 2 degrees for settle_time seconds)
                if abs(pos_error) < math.radians(2.0) and abs(vel_actual) < 0.5:
                    if settle_start is None:
                        settle_start = now
                    elif now - settle_start >= args.settle_time:
                        settled = True
                else:
                    settle_start = None

                if settled:
                    print(f"  Settled at {math.degrees(pos_actual):.1f} deg "
                          f"(error: {math.degrees(pos_error):+.1f} deg)")
                    break

                # Safety timeout
                if now - move_start > 30:
                    print("  TIMEOUT - move took too long!")
                    break

    except KeyboardInterrupt:
        print("\n\nInterrupted!")

    # Stop motor
    print("\nStopping motor...")
    can.send_velocity(0.0)
    time.sleep(0.1)
    can.send_velocity(0.0)
    time.sleep(0.1)
    can.send_nmt(MODE_DISABLED)
    time.sleep(0.1)

    # Final position
    can.send_heartbeat()
    time.sleep(0.1)
    fb = drain_and_get_latest(can)
    if fb:
        print(f"Final position: {math.degrees(fb[0]):.1f} deg")

    print("Done.")
    can.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())
